core/primitives/src/shard_layout.rs

```diff
@@ -83,17 +96,195 @@ pub struct ShardLayoutV1 {
+    }
+
+    pub fn new_end(start: &str) -> Self {
+        Self { start: AccountBoundary::Middle(start.parse().unwrap()), end: AccountBoundary::End }
+    }
+
+    pub fn new_mid(start: &str, end: &str) -> Self {
+        Self {
+            start: AccountBoundary::Middle(start.parse().unwrap()),
+            end: AccountBoundary::Middle(end.parse().unwrap()),
+        }
+    }
+
+    fn contains(&self, account_id: &AccountId) -> bool {
+        match (&self.start, &self.end) {
+            (AccountBoundary::Start, AccountBoundary::End) => true,
+            (AccountBoundary::Start, AccountBoundary::Middle(end)) => account_id < end,
+            (AccountBoundary::Middle(start), AccountBoundary::End) => account_id >= start,
+            (AccountBoundary::Middle(start), AccountBoundary::Middle(end)) => {
+                account_id >= start && account_id < end
+            }
+            (AccountBoundary::End, _) => panic!("invalid account range"),
+            (_, AccountBoundary::Start) => panic!("invalid account range"),
+        }
+    }
+}
+
+/// A mapping from the shard id to the account range of this shard.
+type ShardsAccountRange = BTreeMap<ShardId, AccountRange>;
+
+fn validate_shards_account_range(
+    shards_account_range: &ShardsAccountRange,
+) -> Result<(), ShardLayoutError> {
+    let err =
+        |reason: &str| ShardLayoutError::InvalidShardsAccountRange { reason: reason.to_string() };
+
+    let values = shards_account_range.values().sorted().collect_vec();
+    println!("{:?}", values);
+
+    let (Some(first), Some(last)) = (values.first(), values.last()) else {
+        return Err(err("account range empty"));
+    };
+    if first.start != AccountBoundary::Start {
+        return Err(err("first account range should start with Start"));
+    }
+    if last.end != AccountBoundary::End {
+        return Err(err("last account range should end with End"));
+    }
+
+    for i in 1..values.len() {
+        let prev = values[i - 1];
+        let curr = values[i];
+        if prev.end != curr.start {
+            return Err(err("account ranges should be contiguous"));
+        }
+    }
+
+    Ok(())
+}
+
+/// Making the shard ids non-contiguous.
+#[derive(serde::Serialize, serde::Deserialize, Clone, Debug, PartialEq, Eq)]
+pub struct ShardLayoutV2 {
+    /// A mapping from the shard id to the account range of this shard.
+    shards_account_range: ShardsAccountRange,
+
+    /// A mapping from the parent shard to child shards. Maps shards from the
+    /// previous shard layout to shards that they split to in this shard layout.
+    shards_split_map: Option<ShardsSplitMapV2>,
+    /// A mapping from the child shard to the parent shard. Maps shards in this
+    /// shard layout to their parent shards.
+    shards_parent_map: Option<ShardsParentMapV2>,
+
+    /// Version of the shard layout, this is useful for uniquely identify the shard layout
+    version: ShardVersion,
```

Comment:
I hope with LayoutV2, we would finally increment version by 1?

Reply:
There are a few options here:
1) bump the version and do a migration (since state and flat state are keyed by shard uid)
2) bump the version and add an override on storage layer access, along the lines of `if version >= 3 { version = 3 }` 
3) do not bump the version

The third one doesn't require any extra changes and for that reason it's my favourite. We can just stick to version 3 until we need to make a real change and then we can figure it out. WDYT?

Reply:
no bump ➕